### PR TITLE
[opt] preallocate buffer for evalsha in `Script`

### PR DESCRIPTION
**Pull request by framlog**
a.t.

### CODE REVIEW

File: redis/src/script.rs [hunk 1]
```diff
@@ -213,6 +213,15 @@ impl<'a> ScriptInvocation<'a> {
...
+        let args_len = 3 + self.keys.len() + self.args.len();
+        let buf_len = self
+            .keys
+            .iter()
+            .chain(self.args.iter())
+            .fold(0, |acc, e| acc + e.len())
+            + 7 /* "EVALSHA".len() */
+            + self.script.hash.len();
+        let mut cmd = Cmd::with_capacity(args_len, buf_len);
         let mut cmd = cmd("EVALSHA");
```
Comment (nihohit) [inline redis/src/script.rs:225]: this overwrites the previous `cmd`, and you've essentially done nothing. you'll need to replace this with `arg`.

Comment (framlog) [reply redis/src/script.rs:225]: 0.0..

Sorry, it's a mistake...

File: redis/src/script.rs [hunk 2]
```diff
@@ -213,7 +213,15 @@ impl<'a> ScriptInvocation<'a> {
...
+        let args_len = 3 + self.keys.len() + self.args.len();
+        let buf_len = self
+            .keys
+            .iter()
+            .chain(self.args.iter())
+            .fold(0, |acc, e| acc + e.len())
+            + 7 /* "EVALSHA".len() */
+            + self.script.hash.len();
+        let mut cmd = Cmd::with_capacity(args_len, buf_len);
         cmd.arg(self.script.hash.as_bytes())
```
Comment (nihohit) [inline redis/src/script.rs:225]: now you didn't set the "EVALSHA" argument, and this will fail. 😸

Comment (framlog) [reply redis/src/script.rs:225]: 🧎

I'll fix that.

File: redis/src/script.rs [hunk 3]
```diff
@@ -212,12 +212,44 @@ impl<'a> ScriptInvocation<'a> {
...
 
+    fn estimate_buflen(&self) -> usize {
+        self
+            .keys
+            .iter()
+            .chain(self.args.iter())
+            .fold(0, |acc, e| acc + e.len())
+            + 7 /* "EVALSHA".len() */
+            + self.script.hash.len()
+            + 1 /* Require a little more since rustc seems to reallocate vecs when length matches exactly. */
```
Comment (nihohit) [inline redis/src/script.rs:223]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=26ff6d3d5ddfaf6ad99aeb0c11c20679

I don't see this happening. Can you please explain?

Comment (framlog) [reply redis/src/script.rs:223]: weird...

Comment (framlog) [reply redis/src/script.rs]: I found the real reason is the length of keys.

File: redis/src/cmd.rs [hunk 4]
```diff
@@ -330,6 +330,11 @@ impl Cmd {
         }
     }
 
+    /// Get the capacities for the internal buffers.
+    pub fn capacity(&self) -> (usize, usize) {
```
Comment (nihohit) [inline redis/src/cmd.rs:334]: please change to pub(crate) and add #[cfg(test)]

**Review by nihohit — APPROVED** [review]
Thanks for the contribution! It was a bit of a process, but I think this will be a good perf boost :)

**Pull request merged**